### PR TITLE
Move eligible column to legacy promotions

### PR DESCRIPTION
**Pull request by mamhoff**
## Summary

The `eligible` column on `spree_adjustments` is only ever used in the legacy promotion system. It should live there. 

This simplifies the API of Solidus core quite a bit, as views and serializers do not need to think about whether an adjustment is eligible (tax adjustments are always eligible, for example). SolidusFriendlyPromotions never creates ineligible adjustments. 

The idea is: If an adjustment is there, it's real and can be used and reckoned with. 

Because it works, and this is code that is slated for deprecation, I've not added Deface, but instead overridden all partials that this affects. Particularly mailer partials can't even be targeted with Deface (no HTML, no Deface). 

This is probably the one moment we can move this column out of core.

### CODE REVIEW

**Review by kennyadsl — COMMENTED** [review]
I'm not sure about this one. I'd agree this is a good approach if we don't have to override so many things. 

I'd love to have the simplicity of the API without the `eligible` thing, but probably I'd evaluate keeping it (and noopifying it for the new promotion system). We could deprecate it later, after we extract the legacy promotion system out of the core, so we could have a clean interface in the next major.

I'm a bit battled on this one to be honest, let me know your thoughts on this approach.

File: legacy_promotions/db/migrate/20240622113334_move_adjustment_eligible_to_legacy_promotions.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+class MoveAdjustmentEligibleToLegacyPromotions < ActiveRecord::Migration[7.0]
+  def up
+    unless column_exists?(:spree_adjustments, :eligible)
+      add_column(:spree_adjustments, :eligible, :boolean, default: true)
```
Comment (tvdeyen) [inline legacy_promotions/db/migrate/20240622113334_move_adjustment_eligible_to_legacy_promotions.rb:4]: Now with Rails 7.0 as schema version you can make use of `if_not_exists: true` here

File: core/app/models/spree/adjustment.rb [hunk 2]
```diff
@@ -4,17 +4,8 @@ module Spree
   # Adjustments represent a change to the +item_total+ of an Order. Each
   # adjustment has an +amount+ that can be either positive or negative.
   #
-  # Adjustments can be "opened" or "closed". Once an adjustment is closed, it
+  # Adjustments can be "unfinalizes" or "finalized". Once an adjustment is finalized, it
```
Comment (tvdeyen) [inline core/app/models/spree/adjustment.rb:7]: "unfinalized"?

**Review by tvdeyen — APPROVED** [review]
I do not see an issue with this at all. As long as we release legacy promotions with the next major of solidus everything will behave like before. The `.eligible` scope is still present and does the same as before. Even the template changes wouldn't be necessary imo, but I think it is the right thing to do.

@kennyadsl can you explain what potential issues this change might have?

**Comment by kennyadsl** [comment]
> can you explain what potential issues this change might have?

The only issue I had in mind is that there is a lot of duplication. In the initial phase, we "kinda" need to maintain both versions of the promotion system, and we need to recall to update all the pieces that we are overriding with this PRs on the legacy promotion system.

**Comment by tvdeyen** [comment]
> > can you explain what potential issues this change might have?
> 
> The only issue I had in mind is that there is a lot of duplication. In the initial phase, we "kinda" need to maintain both versions of the promotion system, and we need to recall to update all the pieces that we are overriding with this PRs on the legacy promotion system.

Do we plan to evolve the legacy promotions system or just "keep the lights on"? I hope we will move forward with the new promotion system and keep the legacy one "as is", only fixing security issues and such. Not even fixing bugs (at least not existing ones), because this is the whole part of the deal, right? The legacy system is so flawed (and eligible Adjustments is part of that) we do not even want to see it ever again. At least for me :)

What kind of maintenance do you have in mind?

File: core/spec/models/spree/adjustment_spec.rb [hunk 3]
```diff
@@ -18,6 +18,12 @@
     end
   end
 
+  describe ".eligible", :silence_deprecations do
```
[the remark this answers is not in the document]
Comment (mamhoff) [reply core/spec/models/spree/adjustment_spec.rb:21]: Fixed, thank you!

File: core/app/models/spree/adjustment.rb (hunk 2 above)
Comment (mamhoff) [reply core/app/models/spree/adjustment.rb]: Yes

File: legacy_promotions/db/migrate/20240622113334_move_adjustment_eligible_to_legacy_promotions.rb (hunk 1 above)
Comment (mamhoff) [reply legacy_promotions/db/migrate/20240622113334_move_adjustment_eligible_to_legacy_promotions.rb:4]: Or I can just leave it like this, I don't think there's a big advantage to it.

**Comment by mamhoff** [comment]
So I like to do things fully and not leave unfinished business behind, and the `eligible` column is - to me - unfinished business. I agree it looks like a lot to maintain, but then I don't think it will be: 
For people still using the legacy promotion system, nothing changes (they won't even get the deprecation warnings, because the deprecation is associated to the method object, not the method name!). For people using no promotions or the upcoming `solidus_promotions`, we reward them with a simpler API and less baggage. 

If we're concerned about backwards compatibility, I could make this less dangerous by keeping the implementation and the column, but deprecating it, and removing the column and the implementation with the next major version. I would still keep the partial changes though, so we don't have the baggage of lots of `.eligible`  calls in our templates.

**Review by jarednorman — APPROVED** [review]
I think I agree with @mamhoff on this one.

**Comment by tvdeyen** [comment]
@kennyadsl still have concerns or can we merge?

**Comment by kennyadsl** [comment]
I was sure I answered here at some point, maybe I forgot to press the "Comment" button, as usual. 😄 

I'm totally fine moving forward with this, feel free to merge.

**Comment by MadelineCollier** [comment]
Hey I think we are seeing some consistent spec failures relating to the new deprecation warnings:

```
65) Promotions should display the reimbursements table
      Failure/Error: let!(:reimbursement) { create(:reimbursement) }
      
      ActiveSupport::DeprecationException:
        DEPRECATION WARNING: eligible is deprecated and will be removed from Solidus 5.0 (called from weighted_order_adjustment_amount at /home/circleci/solidus/core/app/models/spree/calculator/returns/default_refund_amount.rb:17)
        
59) Spree::Admin::ReimbursementsController#perform a Spree::Core::GatewayError is raised sets an error message with the correct text
      Failure/Error: let(:reimbursement) { create(:reimbursement) }
      
      ActiveSupport::DeprecationException:
        DEPRECATION WARNING: eligible is deprecated and will be removed from Solidus 5.0 (called from weighted_order_adjustment_amount at /home/circleci/solidus/core/app/models/spree/calculator/returns/default_refund_amount.rb:17)
        
...etc
```        
        
Screenshot 2024-09-02 at 2 15 11 PM

@mamhoff do you know why we'd be seeing that on main but these were never surfaced on the CI runs against branch?

**Comment by mamhoff** [comment]
Yes, because https://github.com/solidusio/solidus/pull/5813 was merged between the opening the PR and merging it. I'll go and fix the warnings.

**Pull request merged**